File: Packages/RallyCore/Sources/RallyCore/Models/ContentItem.swift
import Foundation

/// Represents a content item in the year-round engagement feed.
public struct ContentItem: Identifiable, Codable, Hashable, Sendable {
    public let id: String
    public let schoolID: String
    public let type: ContentType
    public let title: String
    public let body: String?
    public let imageURL: URL?
    public let externalURL: URL?
    public let author: String?
    public let publishedAt: Date
    public let tags: [String]
    public let sponsorID: String?
    public let engagementData: ContentEngagement?

    public init(
        id: String,
        schoolID: String,
        type: ContentType,
        title: String,
        body: String? = nil,
        imageURL: URL? = nil,
        externalURL: URL? = nil,
        author: String? = nil,
        publishedAt: Date = .now,
        tags: [String] = [],
        sponsorID: String? = nil,
        engagementData: ContentEngagement? = nil
    ) {
        self.id = id
        self.schoolID = schoolID
        self.type = type
        self.title = title
        self.body = body
        self.imageURL = imageURL
        self.externalURL = externalURL
        self.author = author
        self.publishedAt = publishedAt
        self.tags = tags
        self.sponsorID = sponsorID
        self.engagementData = engagementData
    }
}

public enum ContentType: String, Codable, Hashable, Sendable {
    case article
    case poll
    case countdown
    case challenge
    case highlight
    case announcement
}

public struct ContentEngagement: Codable, Hashable, Sendable {
    public let likes: Int
    public let comments: Int
    public let shares: Int
    public let pointsValue: Int?

    public init(likes: Int = 0, comments: Int = 0, shares: Int = 0, pointsValue: Int? = nil) {
        self.likes = likes
        self.comments = comments
        self.shares = shares
        self.pointsValue = pointsValue
    }
}

File: Android/core/src/main/java/com/vanwagner/rally/core/model/ContentItem.kt
package com.vanwagner.rally.core.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Represents a content item in the year-round engagement feed.
 */
@Serializable
data class ContentItem(
    val id: String,
    val schoolID: String,
    val type: ContentType,
    val title: String,
    val body: String? = null,
    val imageURL: String? = null,
    val externalURL: String? = null,
    val author: String? = null,
    val publishedAt: Long = System.currentTimeMillis(),
    val tags: List<String> = emptyList(),
    val sponsorID: String? = null,
    val engagementData: ContentEngagement? = null
)

/**
 * Types of content that can appear in the feed.
 */
@Serializable
enum class ContentType {
    @SerialName("article") ARTICLE,
    @SerialName("poll") POLL,
    @SerialName("countdown") COUNTDOWN,
    @SerialName("challenge") CHALLENGE,
    @SerialName("highlight") HIGHLIGHT,
    @SerialName("announcement") ANNOUNCEMENT;

    val displayName: String
        get() = name.lowercase().replaceFirstChar { it.uppercase() }
}

/**
 * Engagement metrics for a content item.
 */
@Serializable
data class ContentEngagement(
    val likes: Int = 0,
    val comments: Int = 0,
    val shares: Int = 0,
    val pointsValue: Int? = null
)
